File: scripts/verify-med-with-solver.mjs
// Run from web-gui: npm run verify:med:solver -- /path/to/solver [julia]
// Uses a read-only solver checkout, its production weak-space functions and
// the existing test fixture; no numerical solver implementation is copied.
import assert from "node:assert/strict";
import {mkdtemp,writeFile,readFile} from "node:fs/promises";
import {tmpdir} from "node:os";
import {join,resolve} from "node:path";
import {fileURLToPath} from "node:url";
import {spawnSync} from "node:child_process";
import {createServer} from "vite";
import {analyzeMed} from "../src/services/medAnalysisService.js";
import {medContactCases,medModel,medBox} from "../test/fixtures/medContactCases.js";

const [solverRoot,julia="julia"]=process.argv.slice(2);
if(!solverRoot) throw new Error("Usage: npm run verify:med:solver -- <solver-root> [julia]");
const server=await createServer({server:{middlewareMode:true},appType:"custom"});
let schema,serialize,deserialize;
try {
    ({default:schema}=await server.ssrLoadModule("/src/services/schemas/elements.schema.js"));
    ({serialize,deserialize}=await server.ssrLoadModule("/src/services/model/modelSerializer.js"));
} finally {await server.close();}
const directory=await mkdtemp(join(tmpdir(),"med-solver-"));
// Julia input is emitted only from these checked, generated numerical fixtures.
const literal=value=>value===null ? "nothing" : typeof value==="string"
    ? JSON.stringify(value).replaceAll("$","\\$") : Array.isArray(value)
    ? `[${value.map(literal).join(",")}]` : typeof value==="object"
    ? `Dict(${Object.entries(value).map(([k,v])=>`${literal(k)}=>${literal(v)}`).join(",")})` : String(value);
const cases=[];
const candidates=medContactCases().filter(c=>!c.error);
candidates.push({name:"original-index",model:medModel([medBox(),medBox({rv:0}),medBox({origin:[1,0,0]})]),contacts:1});
candidates.push({name:"rotated-grid",model:medModel([medBox({dp:[2,2,3]}),medBox({dp:[3,3,2],symVi:[[90],[0],[0]],symR0:[[1],[1],[0]]})]),contacts:1});
for(const symmetry of [{symLs:4,symYl:90},{symAs:4,symYa:90}]) {
    candidates.push({name:symmetry.symLs?"local-seams":"axial-seams",model:medModel([medBox(symmetry)]),contacts:4});
}
const warped=[medBox(),medBox({origin:[1,0,0]})];
for(const row of warped) for(const p of row.geo) p[0]+=.2*p[1]*p[2];
candidates.push({name:"warped-contact",model:medModel(warped),contacts:1});
const moving=medModel([medBox(),medBox({origin:[2,0,0],indMove:1})]);
moving.moves=[{angle:[[0,0,0,0],[1,0,0,0]],position:[[0,-1,0,0],[1,10,0,0]]}];
candidates.push({name:"initial-motion",model:moving,contacts:1});
for(const doubleFloat of [false,true]) for(const c of candidates) {
    const model=structuredClone(c.model);model.general.doubleFloat=doubleFloat;
    const result=analyzeMed(model);assert.deepEqual(result.errors,[],c.name);
    const updated=model.elements.map((row,i)=>({...row,med:result.faces.filter(f=>f.block===i+1).length
        ? result.faces.filter(f=>f.block===i+1).sort((a,b)=>a.face-b.face).map(f=>[f.proposed]) : row.med}));
    const path=join(directory,`${c.name}-${doubleFloat?64:32}-kvs.txt`);
    await writeFile(path,JSON.stringify(serialize(updated,schema)));
    const storage=JSON.parse(await readFile(path,"utf8"));
    assert.deepEqual(deserialize(storage,schema).map(r=>r.med),updated.map(r=>r.med));
    cases.push({name:c.name,doubleFloat,general:model.general,kvs:storage,moves:model.moves,contacts:c.contacts});
}
const dataPath=join(directory,"cases.jl");
await writeFile(dataPath,`const MED_CASES = ${literal(cases)}\n`);
console.log(`Control JSON: ${directory}`);
const script=fileURLToPath(new URL("verify-med-with-solver.jl",import.meta.url));
const run=spawnSync(julia,["--startup-file=no",script,resolve(solverRoot),dataPath],{stdio:"inherit"});
if(run.error) throw run.error;
process.exitCode=run.status ?? 1;

File: scripts/verify-med-with-solver.jl
using Test, LinearAlgebra
length(ARGS)==2 || error("Expected solver checkout and generated cases.jl")
const SOLVER_ROOT=abspath(ARGS[1])
include(joinpath(SOLVER_ROOT,"test","support","current_weak_fixture.jl"))
include(abspath(ARGS[2]))

function check_case(m,c)
    T=m.REAL
    task=m.TaskData()
    task.general.mirrorSymmetryX=c["general"]["mirrorSymmetryX"]
    task.general.mirrorSymmetryY=c["general"]["mirrorSymmetryY"]
    for move in c["moves"]
        push!(task.moves,m.Move(angle=permutedims(hcat([T.(row) for row in move["angle"]]...)),
            position=permutedims(hcat([T.(row) for row in move["position"]]...))))
    end
    for row in c["kvs"]
        kv=m.cube(targ=row["targ"],rv=row["rv"]>0,model=row["model"])
        kv.self.geoType=row["geoType"]
        kv.self.geo=T.(reduce(vcat,row["geo"]))
        kv.self.dr=T.(row["dr"]);kv.self.dp=Int.(row["dp"])
        kv.self.rv=T(row["rv"]);kv.self.med=Int.(row["med"])
        kv.self.indMove=row["indMove"]
        sym=row["sym"]
        for k in ("ls","as","ps","kya","kyp")
            setproperty!(kv.self.sym,Symbol(k),Int(sym[k]))
        end
        for k in ("yl","ya","tx")
            setproperty!(kv.self.sym,Symbol(k),T(sym[k]))
        end
        kv.self.sym.vi=T.(sym["vi"]);kv.self.sym.r0=T.(sym["r0"])
        m.update_sym(kv.self.sym,kv.sym)
        m.update_sym(kv.self.sym,kv.sym,zeros(T,3),zeros(T,3))
        @test m.unpack(kv)==0
        push!(task.kvs,kv)
    end
    m.evalCounters(task)
    m.moveRecall(task,0)
    topology=m.buildCurrentTopology(task)
    space,cells=m.weakCurrentSpace(task)
    @test length(cells)==length(topology.cells)
    @test all(isfinite,space.c.nzval)
    @test all(>(0),space.weights)
    # Every free external face must stay insulating in the production topology.
    for (i,cell) in enumerate(cells), f in 1:6
        if cell.boundary[f] && task.kvs[cell.block].self.med[f]==-1
            @test topology.faces[f,i]==0
        end
    end
    insulated=deepcopy(task)
    for kv in insulated.kvs;kv.self.med .= -1;end
    separate=m.weakCurrentSpace(insulated)[1]
    if c["contacts"]>0
        @test size(space.c,1)<size(separate.c,1)
        # Destroy exactly one side of a valid pointer: production must reject it.
        broken=deepcopy(task)
        pointer=findfirst(kv->any(>(0),kv.self.med),broken.kvs)
        if !isnothing(pointer)
            f=findfirst(>(0),broken.kvs[pointer].self.med)
            broken.kvs[pointer].self.med[f]=-1
            @test_throws ErrorException m.weakCurrentSpace(broken)
        end
    else
        @test size(space.c)==size(separate.c)
    end
end

@testset "web-gui MED / existing solver weak space" begin
    for T in (Float32,Float64)
        m=CurrentWeakTests.fixture(T)
        Base.include(m,joinpath(SOLVER_ROOT,"src","base","04_symmetry.jl"))
        Base.include(m,joinpath(SOLVER_ROOT,"src","base","04_approx.jl"))
        Base.include(m,joinpath(SOLVER_ROOT,"src","task","02_moves.jl"))
        for c in MED_CASES
            c["doubleFloat"]==(T===Float64) || continue
            @testset "$(c["name"]) $T" begin
                Base.invokelatest(check_case,m,c)
            end
        end
    end
end
